service/history/workflow/mutable_state_impl.go

```diff
@@ -2130,6 +2133,66 @@ func (ms *MutableStateImpl) addResetPointFromCompletion(
 	return true
 }
 
+// processBuildIdRedirect validates build id for the task being dispatched and applies possible redirect to
+// mutable state.
+// If a valid redirect is happening, assigned build id of the wf will be updated and all scheduled but not
+// started tasks will be rescheduled to be put on the matching queue of the right build id.
+// If the given versioning stamp and redirect info is not valid based on the wf's assigned build id
+// InvalidDispatchBuildId error will be returned.
+func (ms *MutableStateImpl) processBuildIdRedirect(
+	startingTaskScheduledEventId int64,
+	versioningStamp *commonpb.WorkerVersionStamp,
+	redirectInfo *taskqueue.BuildIdRedirectInfo,
+) error {
+	if !versioningStamp.GetUseVersioning() || versioningStamp.GetBuildId() != ms.GetAssignedBuildId() {
+		// dispatch build id is the same as wf assigned build id, hence noop.
+		return nil
+	}
+
+	if redirectInfo == nil || redirectInfo.GetAssignedBuildId() != ms.GetAssignedBuildId() {
+		// no redirect or a redirect based on a wrong assinged build id is reported. this must be a task
+		// backlogged on an old build id. rejecting this task, there should be another task scheduled on
+		// the right build id.
+		return serviceerrors.NewInvalidDispatchBuildId()
+	}
+
+	// redirect is valid, updating mutable state and re-scheduling pending tasks which are not started yet.
+	newBuildId := versioningStamp.GetBuildId()
+	err := ms.UpdateBuildIdAssignment(newBuildId)
+	if err != nil {
+		return err
+	}
+
+	if ms.HasPendingWorkflowTask() && !ms.HasStartedWorkflowTask() &&
+		ms.GetPendingWorkflowTask().ScheduledEventID != startingTaskScheduledEventId &&
+		// speculative tasks do not go to matching
+		ms.GetPendingWorkflowTask().Type != enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE {
+		// sticky queue should be cleared by UpdateBuildIdAssignment already, so the following call only generates
+		// a WorkflowTask, not a WorkflowTaskTimeoutTask.
+		err = ms.taskGenerator.GenerateScheduleWorkflowTaskTasks(ms.GetPendingWorkflowTask().ScheduledEventID)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, ai := range ms.GetPendingActivityInfos() {
+		if ai.ScheduledEventId == startingTaskScheduledEventId ||
+			// activity already started
+			ai.StartedEventId != common.EmptyEventID ||
+			// activity does not depend on wf build id
+			ai.GetUseWorkflowBuildId() == nil {
+			continue
+		}
+		// we only need to resend the activities to matching, no need to update timer tasks.
+		err = ms.taskGenerator.GenerateActivityTasks(&historypb.HistoryEvent{EventId: ai.ScheduledEventId})
```

Comment:
Not sure we need to generate all of these tasks.
What if the activity is backing off?
What if the task has the right build ID?

Reply:
> What if the activity is backing off? We don't want to retrigger it prematurely.
It seems we do not track activity timer status in mutable state as of now. So preventing new activity tasks during backoff period would need some more work. I'm not sure it's necessarily a bad thing to trigger activity prematurely because the new build is likely fixing the error you're retrying for and it may be even desirable to skip the backoff.

> What if the task has the right build ID?
You mean if the activity already has a task queued in the new build ID? this can only happen when we are redirecting to a build ID that the wf has used previously. It should be a rare case, is not worth optimizing for IMHO. Also, we don't have visibility on all the build IDs that a particular activity is scheduled on anyways.

Reply:
I'd accept the code as-is but I'd want to make sure we have follow ups.
I think we should respect the retry policy and try to not create duplicate tasks as much as possible, all we need is to record the last attempt failure time and to calculate the next attempt time based on the retry policy.
AFAIU duplicates may end up increasing the frequency of attempts.

Reply:
Sounds good. Created a task for it to be addressed before Public Preview. Also added a TODO in that part of the code.